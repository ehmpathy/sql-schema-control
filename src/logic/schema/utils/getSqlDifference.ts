import { diffStringsUnified } from 'jest-diff';

export const getSqlDifference = ({
  oldSql,
  newSql,
}: {
  oldSql: string;
  newSql: string;
}) => {
  // check strings are not equivalent
  if (oldSql === newSql) return null; // i.e., no difference

  // compute the differences
  const sqlDiffString = diffStringsUnified(oldSql.trim(), newSql.trim());

  // return the diff string
  return sqlDiffString;
};
