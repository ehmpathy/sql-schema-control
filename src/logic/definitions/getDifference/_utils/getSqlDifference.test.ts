import { getSqlDifference } from './getSqlDifference';

describe('getSqlDifference', () => {
  it('should find that a colored diff of changes if sql is different', async () => {
    const oldSql = `
CREATE TABLE IF NOT EXISTS some_table (
id              INT(11) PRIMARY KEY AUTO_INCREMENT,
created_at      DATETIME DEFAULT CURRENT_TIMESTAMP,
) ENGINE = InnoDB;
    `;
    const newSql = `
CREATE TABLE IF NOT EXISTS some_table (
id              BIGINT PRIMARY KEY AUTO_INCREMENT,
created_at      DATETIME(3) DEFAULT CURRENT_TIMESTAMP
) ENGINE = InnoDB;
    `;

    // get the diff
    const result = await getSqlDifference({ oldSql, newSql });
    expect(typeof result).toEqual('string');
    expect(result).toMatchSnapshot(); // result is purely visual, so log an example of it
  });

});
