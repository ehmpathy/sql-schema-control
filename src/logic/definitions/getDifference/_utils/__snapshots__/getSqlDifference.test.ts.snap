// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSqlDifference should find that a colored diff of changes if sql is different 1`] = `
"[90m[39m
[90mCREATE TABLE IF NOT EXISTS some_table ([39m
[90m[39m[31mid              INT(11) PRIMARY KEY AUTO_INCREMENT,[39m
[31mcreated_at      DATETIME DEFAULT CURRENT_TIMESTAMP,[39m
[31m[39m[32mid              BIGINT PRIMARY KEY AUTO_INCREMENT,[39m
[32mcreated_at      DATETIME(3) DEFAULT CURRENT_TIMESTAMP[39m
[32m[39m[90m) ENGINE = InnoDB;[39m
[90m[39m"
`;
